src/norm/pool.nim

```diff
@@ -0,0 +1,106 @@
+import std/locks
+
+import sqlite, postgres
+
+
+type 
+  Conn = sqlite.DbConn | postgres.DbConn
+  Pool*[T: Conn] = ref object 
+    defaultSize: Natural
+    conns: seq[T]
+    poolExhaustedPolicy: PoolExhaustedPolicy
+    lock: Lock
+  PoolExhaustedError* = object of CatchableError
+  PoolExhaustedPolicy* = enum
+    pepRaise
+    pepExtend
+
+
+func newPool*[T: Conn](defaultSize: Positive, poolExhaustedPolicy = pepRaise): Pool[T] =
+  ##[ Create a connection pool of the given size.
+
+  ``poolExhaustedPolicy`` defines how the pool reacts when a connection is requested but the pool has no connection available:
+
+    - ``pepRaise`` (default) means throw ``PoolExhaustedError``
+    - ``pepExtend`` means throw “add another connection to the pool.”
+  ]##
+
+  result = Pool[T](defaultSize: defaultSize, conns: newSeq[T](defaultSize), poolExhaustedPolicy: poolExhaustedPolicy)
+
+  initLock(result.lock)
+
+  for conn in result.conns.mitems:
+    conn =
+      when T is sqlite.DbConn:
```

Comment:
@PhilippMDoerner I've managed to make the implementation universal by making Pool type generic and using this kind of ugly hack. I'm not super happy with this part but it does the job done and do so without any runtime penalty.

Reply:
Given that you do this in two different places it might be worth considering moving this "when"-ified getDb call into a private proc that deals with this

Reply:
Actually, wait!
We might want to contemplate **not** using getDb() at all.

A significant benefit of allowing multiple, user defined pools is the fact that you might want to connect to multiple databases.
If we use getDb, that is not the case, because you'll always be using the same environment variables.

We might want to contemplate allowing the user to pass a proc that generates DbConn instead and use that instead of getDb! getDb could be the default proc for such a parameter though.

Reply:
Just for reference, I consider my comment here as resolved as you implemented it as the user passing in a connection-creation proc.